From 3b6b4eba9cb531e357da7a4f4b26fc779b89e495 Mon Sep 17 00:00:00 2001
From: Jai Luthra <j-luthra@ti.com>
Date: Fri, 23 Jun 2023 18:14:43 +0530
Subject: [PATCH 081/143] media: ti: j721e-csi2rx: Fix buffer cleanup sequence

Ensure that DMA is terminated (and any active buffers are marked as done
in the dma channel callback) before marking the remaining buffers in the
queue as done.

Without this fix there were rare errors reported inside dma callback if
the active buffer was already returned to VB2 and thus got freed twice.

Fixes: 63bc1fbc5c85 ("media: ti: j721e-csi2rx: Fix stream stop sequence")
Signed-off-by: Jai Luthra <j-luthra@ti.com>
---
 .../platform/ti/j721e-csi2rx/j721e-csi2rx.c   | 22 +++++++++----------
 1 file changed, 11 insertions(+), 11 deletions(-)

diff --git a/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c b/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
index af4569fe902f..9851385e30ec 100644
--- a/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
+++ b/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
@@ -743,20 +743,13 @@ static void ti_csi2rx_cleanup_buffers(struct ti_csi2rx_ctx *ctx,
 				      enum vb2_buffer_state buf_state)
 {
 	struct ti_csi2rx_dma *dma = &ctx->dma;
-	struct ti_csi2rx_buffer *buf = NULL, *tmp, *curr;
+	struct ti_csi2rx_buffer *buf = NULL, *tmp;
 	enum ti_csi2rx_dma_state state;
 	unsigned long flags;
 	int ret;
 
 	spin_lock_irqsave(&dma->lock, flags);
-	list_for_each_entry_safe(buf, tmp, &ctx->dma.queue, list) {
-		list_del(&buf->list);
-		vb2_buffer_done(&buf->vb.vb2_buf, buf_state);
-	}
-
-	curr = ctx->dma.curr;
 	state = ctx->dma.state;
-	dma->curr = NULL;
 	dma->state = TI_CSI2RX_DMA_STOPPED;
 	spin_unlock_irqrestore(&dma->lock, flags);
 
@@ -765,11 +758,18 @@ static void ti_csi2rx_cleanup_buffers(struct ti_csi2rx_ctx *ctx,
 		if (ret)
 			dev_dbg(ctx->csi->dev,
 				"Failed to drain DMA. Next frame might be bogus\n");
-		dmaengine_terminate_sync(ctx->dma.chan);
 	}
+	dmaengine_terminate_sync(ctx->dma.chan);
 
-	if (curr)
-		vb2_buffer_done(&curr->vb.vb2_buf, buf_state);
+	spin_lock_irqsave(&dma->lock, flags);
+	list_for_each_entry_safe(buf, tmp, &ctx->dma.queue, list) {
+		list_del(&buf->list);
+		vb2_buffer_done(&buf->vb.vb2_buf, buf_state);
+	}
+
+	if (dma->curr)
+		vb2_buffer_done(&dma->curr->vb.vb2_buf, buf_state);
+	spin_unlock_irqrestore(&dma->lock, flags);
 }
 
 static int ti_csi2rx_restart_dma(struct ti_csi2rx_ctx *ctx,
-- 
2.41.0

